server/container/core/src/test/java/org/apache/james/core/MailImplTest.java

```diff
@@ -80,31 +87,173 @@ public void testConstr3() throws MessagingException {
+    }
+
+    @Test
+    public void getAttributeShouldReturnStoredValue() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key = "test";
+        mail.setAttribute(key, value);
+
+        assertThat(mail.getAttribute(key)).isEqualTo(value);
+    }
+
+    @Test
+    public void setAttributeShouldReturnPreviouslyStoredValue() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key = "test";
+        mail.setAttribute(key, value);
+
+        assertThat(mail.setAttribute(key, "newValue")).isEqualTo(value);
+    }
+
+    @Test
+    public void removeAttributeShouldReturnNullWhenAbsent() {
+        Mail mail = new MailImpl();
+
+        assertThat(mail.removeAttribute("test")).isNull();
+    }
+
+    @Test
+    public void removeAttributeShouldReturnPreviouslyStoredValue() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key = "test";
+        mail.setAttribute(key, value);
+
+        assertThat(mail.removeAttribute(key)).isEqualTo(value);
+    }
+
+    @Test
+    public void hasAttributesShouldReturnTrueWhenHasAttribute() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key = "test";
+        mail.setAttribute(key, value);
+
+        assertThat(mail.hasAttributes()).isTrue();
+    }
+
+    @Test
+    public void getAttributeNamesShouldReturnStoredAttributesName() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key = "test";
+        mail.setAttribute(key, value);
+
+        assertThat(mail.getAttributeNames()).containsExactly(key);
+    }
+
+    @Test
+    public void removeAttributeShouldRemoveSpecificAttribute() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key1 = "test1";
+        String key2 = "test2";
+        mail.setAttribute(key1, value);
+        mail.setAttribute(key2, value);
+        mail.removeAttribute(key1);
+
+        assertThat(mail.getAttributeNames()).containsExactly(key2);
+    }
+
+    @Test
+    public void removeAllAttributesShouldClearMailAttributes() {
+        Mail mail = new MailImpl();
+
+        String value = "value";
+        String key1 = "test1";
+        String key2 = "test2";
+        mail.setAttribute(key1, value);
+        mail.setAttribute(key2, value);
+        mail.removeAllAttributes();
+
+        assertThat(mail.getAttributeNames()).isEmpty();
+    }
+
+    protected void helperTestInitialState(Mail mail) {
+        assertThat(mail.hasAttributes()).as("Expecting no initial attributes").isFalse();
+        assertThat(mail.getErrorMessage()).as("Expecting no initial error").isNull();
+        assertThat(mail.getLastUpdated()).as("Expecting initial last update set").isNotNull();
+        try {
+            mail.getRecipients().isEmpty();
+        } catch (NullPointerException e) {
+            // current behavior. *BUT*, shouldn't this method better return with
```

Comment:
Agree. See next commits. 

Upgrade code quality first.

Fix and change behaviour later.